app/src/main/cpp/skyline/common/settings.h

```diff
@@ -3,23 +3,72 @@
 
 #pragma once
 
-#include <common.h>
+#include "language.h"
 
 namespace skyline {
     /**
-     * @brief The Settings class is used to access preferences set in the Kotlin component of Skyline
+     * @brief The Settings class provides a simple interface to access user-defined settings, update values and subscribe callbacks to observe changes.
      */
     class Settings {
+        template<typename T>
+        class Setting {
+            using Callback = std::function<void(const T &)>;
+            std::vector<Callback> callbacks; //!< Callbacks to be called when this setting changes
+            T value;
+
+            /**
+             * @brief Calls all registered callbacks
+             */
+            void OnSettingChanged() {
+                std::for_each(callbacks.begin(), callbacks.end(), [&](const Callback &callback) {
+                    callback(value);
+                });
+            }
+
+          public:
+            /**
+             * @brief Getter of the underlying setting value
+             */
+            const T &operator*() const {
+                return value;
+            }
+
+            /**
+             * @brief Setter of the underlying setting value
+             */
+            void operator=(T newValue) {
+                if (value != newValue) {
+                    value = std::move(newValue);
+                    OnSettingChanged();
+                }
+            }
+
+            /**
+             * @brief Register a callback to run when this setting changes
+             */
+            void AddCallback(Callback callback) {
+                callbacks.push_back(std::move(callback));
+            }
+        };
+
       public:
-        Logger::LogLevel logLevel; //!< The minimum level that logs need to be for them to be printed
-        std::string username; //!< The name set by the user to be supplied to the guest
-        bool operationMode; //!< If the emulated Switch should be handheld or docked
-        bool forceTripleBuffering; //!< If the presentation engine should always triple buffer even if the swapchain supports double buffering
-        bool disableFrameThrottling; //!< Allow the guest to submit frames without any blocking calls
+        // System
+        Setting<bool> isDocked; //!< If the emulated Switch should be handheld or docked
+        Setting<std::string> usernameValue; //!< The name set by the user to be supplied to the guest
+        Setting<language::SystemLanguage> systemLanguage; //!< The system language set by the user
+
+        // Display
+        Setting<bool> forceTripleBuffering; //!< If the presentation engine should always triple buffer even if the swapchain supports double buffering
+        Setting<bool> disableFrameThrottling; //!< Allow the guest to submit frames without any blocking calls
+
+        Settings() {}
```

Comment:
I don't think we want this to be a POD when it's not, I think we should keep the empty ctor.

Reply:
The compiler isn't silly, it won't mark a type as trivial just because of a defaulted ctor, defaulted ctor just doesn't prevent it. In this case it's not trivial cause of the virtual member. Defaulting a ctor otherwise gives exact same behaviour while being more uniform